### ISSUE TITLE
Weird memory usage for 11m vs 160m: same batch size fits in memory...

### ISSUE THREAD

**Issue by alexjc** [mlfoundations/open_lm#35]
I haven't yet investigated the cause of this, but I'd appreciate insights and help.

When running a 160m model (``hidden_dim: 768, n_layers: 12, n_heads: 12, seq_len: 256``) I can fit `batch_size: 48` in memory on a 3090, and it saturates the GPUs nicely.

Then running a 11m model (``hidden_dim: 96, n_layers: 12, n_heads: 12, seq_len: 256``) I can **still only** fit `batch_size: 48` in memory on a 3090.

Things to note:
- The vocab size is the same, the code is exactly the same for both runs.
- I've disabled gradient checkpointing and FSDP given the model sizes.

Would you expect to be able to increase the `batch_size` for smaller models?  If so, there may be a problem somewhere...

I've used other LLM codebases, but not yet familiar with this one, so any insights where to look for problems would be appreciated!

**Comment by sagadre** [comment]
Hi @alexjc thanks for the issue! I’ll try to reproduce on my end, but in the meantime can your please share the command lines you are running?

**Comment by alexjc** [comment]
As well as the `seq_len: 256,` changes to the JSON config, here is the `run_train.sh` script I'm using:

```
#!/bin/bash

export OMP_NUM_THREADS=1
torchrun --nproc-per-node 2 -m open_lm.main --model open_lm_11m \
    --train-data preproc_data/2048-v1/0/shard-{0000000..0000100}.tar --train-num-samples 1000000000 \
    --workers 8 --dataset-resampled --precision amp_bfloat16 --batch-size 55 \
    --log-every-n-steps 100 --grad-clip-norm 1 --data-key txt --lr 3e-4 \
    --warmup 2000 --wd 0.1 --beta2 0.95
```

Binary search results:

- **160m model** reaches `bach_size: 52` without OOM, 53 fails.
- **11m model** reaches `batch_size: 56` without OOM, 57 fails.

I'm going to fetch the latest version of the dependencies and try that before digging too much.

**Comment by sagadre** [comment]
One thing that jumps out is not sure 3090s have bf16 support. May also be worth trying fp32 and fp16 for debugging

**Comment by alexjc** [comment]
I'm pretty sure 3090s do support bf16 but I'll test regular float16!

**Comment by alexjc** [comment]
Using `amp_bf16` instead of `amp_bfloat16` results in the same memory usage.  160m reaches `bs=52`, 11m reaches `bs=56`.

So far I believe it's related to hardware support, and I don't think AMP is a significant factor here... 🤔

**Comment by alexjc** [comment]
OK, so right before entering `train_one_epoch()` the memory usage of GPU 0 is proportional to the number of workers.

With only one worker and one GPU, this is the memory allocated:
```
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Active memory         | 776348 KiB | 776348 KiB | 776348 KiB |      0 B   |
```

With two workers and two GPUs, this is the memory allocated on GPU 0:
```
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |   1514 MiB |   1514 MiB |   2272 MiB | 776210 KiB |
|---------------------------------------------------------------------------|
| Active memory         |   2272 MiB |   2272 MiB |   2272 MiB |   6144 B   |
```

It seems there's a full copy of the model for GPU 1 on GPU 0 too, and another extra copy that likely stores the combined parameters?  After the training function is entered, I presume this will also include activations as well as the gradients...

**Comment by alexjc** [comment]
Setting this DistributedDataParallel parameter according to the docs makes the GPUs carry equal load before entering the `train_one_epoch()`.

```
L465:            ddp_args = {"output_device": device}
```

The OOM remain the same though, but both GPUs now throw exceptions at the same time.

**Comment by alexjc** [comment]
UPDATE: Results above look like the cost of doing business with `DistributedDataParallel`, nothing too out of the ordinary?

The memory usage at peak in `train_one_epoch` is interesting, just before the call to `backward()`:
- **160m model** peak allocated 14.3 G.
- **11m model* peak allocated 9.5 G.

Still tracing it through the rest of the function, that's not where it hands but looks a bit suspicious already...

**Comment by anas-awadalla** [comment]
> I haven't yet investigated the cause of this, but I'd appreciate insights and help.
> 
> When running a 160m model (`hidden_dim: 768, n_layers: 12, n_heads: 12, seq_len: 256`) I can fit `batch_size: 48` in memory on a 3090, and it saturates the GPUs nicely.
> 
> Then running a 11m model (`hidden_dim: 96, n_layers: 12, n_heads: 12, seq_len: 256`) I can **still only** fit `batch_size: 48` in memory on a 3090.
> 
> Things to note:
> 
> * The vocab size is the same, the code is exactly the same for both runs.
> * I've disabled gradient checkpointing and FSDP given the model sizes.
> * I'm just using two GPUs for testing, memory runs out on the first process.
> 
> Would you expect to be able to increase the `batch_size` for smaller models? If so, there may be a problem somewhere...
> 
> I've used other LLM codebases, but not yet familiar with this one, so any insights where to look for problems would be appreciated!

If it is helpful I also saw similar behavior. My runs where with seq length 2048 on 4 A100 nodes.

**Comment by mitchellnw** [comment]
This is quite weird, thanks a lot for documenting that this is an issue. Just curious, does the behavior go away with `--grad-checkpointing`?

**Comment by anas-awadalla** [comment]
Not for me. Used grad checkpointing for both runs.

**Comment by alexjc** [comment]
Yes, it helps I'm not chasing a white rabbit! 🐇🕳

For `float32`, peak allocated `10.3G` for 11m model before backward(), compared to `9.5G` with `bfloat16`.

- The relative memory usage for 16-bit and 32-bit don't make sense to me.
- The relative memory usage for 160m vs 11m parameters also doesn't.

@mitchellnw Let me check! To clarify what I'm looking for, I'd expect 12x to 14x batch sizes to fit for 11m vs 160m?

**Comment by alexjc** [comment]
Some progress, isolating the problem to a single GPU single worker helps. The problems are there also.

- **BUG**: Models with single GPU/workers don't seem to be correctly cast to fp16. With 160m model, I can fit `batch_size: 61` with fp32, and only `batch_size: 58` with fp16. The difference of 3 is probably due to the extra memory of converting output to float() "like LLaMa", which requires an extra buffer for fp16.
- **OK**: I removed `autocast()` and that doesn't seem to be the problem. I also integrated the PyTorch built-in `scaled_dot_product_attention()` instead of xformers and that doesn't change overall memory usage much either.

**Comment by mitchellnw** [comment]
> @mitchellnw Let me check! To clarify what I'm looking for, I'd expect 8x / 12x / 14x batch sizes to fit for 11m vs 160m?

Yes totally. Sorry about this issue and thanks for all the debugging... Sorry I don't really have anything helpful to add at this point.

**Comment by alexjc** [comment]
@mitchellnw OK, thanks. The good news is that it's easy to isolate and reproduce!

Having removed the autocast and fixed the raw fp16 problem, I made a chart of the memory usage before the backward() pass.  For 11m it's perfectly linear, but the slope of the line is not what you'd expect relative to 160m.

<img width="897" alt="Screenshot 2023-10-16 at 10 22 05" src="https://github.com/mlfoundations/open_lm/assets/445208/3b36e8a7-8ed1-486b-921f-7e0d753d3d94">

I'm out of ideas (and time) for now, hopefully there'll be a ready-formed insight that appears!
